### ISSUE TITLE
Assert failure: (ptrdiff_t)generation_free_list_space (gen) >= 0

### ISSUE THREAD

**Issue by echesakovMSFT** [dotnet/runtime#57956]
```
set COMPlus_TieredCompilation=0
set COMPlus_JitStress=2
set COMPlus_JitStressRegs=2
```

```
    GC\Features\HeapExpansion\bestfit\bestfit.cmd [FAIL]
      
      Assert failure(PID 3468 [0x00000d8c], Thread: 13884 [0x363c]): (ptrdiff_t)generation_free_list_space (gen) >= 0
      
      CORECLR! WKS::gc_heap::soh_try_fit + 0x380 (0x00007fff`efc89a08)
      CORECLR! WKS::gc_heap::allocate_soh + 0x334 (0x00007fff`efc651a4)
      CORECLR! WKS::gc_heap::try_allocate_more_space + 0x370 (0x00007fff`efc8b760)
      CORECLR! WKS::gc_heap::allocate_more_space + 0x34 (0x00007fff`efc64e4c)
      CORECLR! WKS::GCHeap::Alloc + 0x12C (0x00007fff`efc5dd4c)
      CORECLR! Alloc + 0x164 (0x00007fff`efaf4c6c)
      CORECLR! AllocateSzArray + 0x178 (0x00007fff`efaf5b40)
      CORECLR! JIT_NewArr1 + 0x16C (0x00007fff`efb163ac)
      CORECLR! JIT_NewArr1VC_MP_FastPortable + 0x194 (0x00007fff`efb208c4)
      <no module>! <no symbol> + 0x0 (0x00007fff`90261100)
          File: D:\workspace\_work\1\s\src\coreclr\gc\gc.cpp Line: 15585
          Image: D:\h\w\9DBE08FB\p\corerun.exe
      
      
      Return code:      1
      Raw output file:      D:\h\w\9DBE08FB\w\ACD10910\uploads\Reports\GC.Features\HeapExpansion\bestfit\bestfit.output.txt
      Raw output:
      BEGIN EXECUTION
       "D:\h\w\9DBE08FB\p\corerun.exe" -p "System.Reflection.Metadata.MetadataUpdater.IsSupported=false" bestfit.dll 1 1000 50000 
      Using random seed: -528163915
      Took 7674 iterations to reach steady state
      Expected: 100
      Actual: -1073740286
      END EXECUTION - FAILED
      FAILED
      Test Harness Exitcode is : 1
      To run the test:
      > set CORE_ROOT=D:\h\w\9DBE08FB\p
      > D:\h\w\9DBE08FB\w\ACD10910\e\GC\Features\HeapExpansion\bestfit\bestfit.cmd
      Expected: True
      Actual:   False
      Stack Trace:
           at GC_Features._HeapExpansion_bestfit_bestfit_._HeapExpansion_bestfit_bestfit_cmd()
      Output:
        
        Assert failure(PID 3468 [0x00000d8c], Thread: 13884 [0x363c]): (ptrdiff_t)generation_free_list_space (gen) >= 0
        
        CORECLR! WKS::gc_heap::soh_try_fit + 0x380 (0x00007fff`efc89a08)
        CORECLR! WKS::gc_heap::allocate_soh + 0x334 (0x00007fff`efc651a4)
        CORECLR! WKS::gc_heap::try_allocate_more_space + 0x370 (0x00007fff`efc8b760)
        CORECLR! WKS::gc_heap::allocate_more_space + 0x34 (0x00007fff`efc64e4c)
        CORECLR! WKS::GCHeap::Alloc + 0x12C (0x00007fff`efc5dd4c)
        CORECLR! Alloc + 0x164 (0x00007fff`efaf4c6c)
        CORECLR! AllocateSzArray + 0x178 (0x00007fff`efaf5b40)
        CORECLR! JIT_NewArr1 + 0x16C (0x00007fff`efb163ac)
        CORECLR! JIT_NewArr1VC_MP_FastPortable + 0x194 (0x00007fff`efb208c4)
        <no module>! <no symbol> + 0x0 (0x00007fff`90261100)
            File: D:\workspace\_work\1\s\src\coreclr\gc\gc.cpp Line: 15585
            Image: D:\h\w\9DBE08FB\p\corerun.exe
        
        
        Return code:      1
        Raw output file:      D:\h\w\9DBE08FB\w\ACD10910\uploads\Reports\GC.Features\HeapExpansion\bestfit\bestfit.output.txt
        Raw output:
        BEGIN EXECUTION
         "D:\h\w\9DBE08FB\p\corerun.exe" -p "System.Reflection.Metadata.MetadataUpdater.IsSupported=false" bestfit.dll 1 1000 50000 
        Using random seed: -528163915
        Took 7674 iterations to reach steady state
        Expected: 100
        Actual: -1073740286
        END EXECUTION - FAILED
        FAILED
        Test Harness Exitcode is : 1
        To run the test:
        > set CORE_ROOT=D:\h\w\9DBE08FB\p
        > D:\h\w\9DBE08FB\w\ACD10910\e\GC\Features\HeapExpansion\bestfit\bestfit.cmd
```

Failed in [runtime-coreclr jitstress2-jitstressregs #20210821.1](https://dev.azure.com/dnceng/public/_build/results?buildId=1310171) - [console.log](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-25983977d49647c5b6/PayloadGroup0/1/console.303cf47e.log?sv=2019-07-07&se=2021-09-10T18%3A09%3A12Z&sr=c&sp=rl&sig=JCtETiVxbUYBHdh5O7v0y6WErOUCqf2RTM35v5405QY%3D)

**Comment by mangod9** [comment]
Hi @echesakovMSFT , do you happen to know if this is a regression?

**Comment by echesakovMSFT** [comment]
@mangod9 I see that the test failed only once (in the test run I quoted), but since it failed with assert I assumed it might be worth looked at.

**Comment by mangod9** [comment]
@cshung could this be related to any of your changes recently?

**Comment by Maoni0** [comment]
@PeterSolMS checked in the assert recently but that most likely just exposes a problem that was there before. @PeterSolMS, could you please take a look?

**Comment by mangod9** [comment]
@PeterSolMS assume doesnt need a backport to 6 so can be closed based on your PR?

**Comment by PeterSolMS** [comment]
Have we observed any failures since I merged this a week ago? If not, I think we can close it (and re-open if needed).

I don't think we need to backport to 6. 

**Rationale:**
The assert firing shows that our bookkeeping for free list size becomes slightly wrong due to the race condition. In retail builds, there would be no assert, of course, but our decision to compact gen 0 may be slightly off in some cases - we would compact more often than we really intended to. As we mostly compact gen 0 anyways, this seems pretty harmless, and we would reset the generation_free_list_space in the next GC anyway, so at most once gen 0 GC would be affected anytime this happens.
